staging/src/k8s.io/apiserver/pkg/server/filters/reqmgmt_test.go

```diff
@@ -0,0 +1,409 @@
+				Exempt:                   false,
+				GlobalDefault:            false,
+				HandSize:                 int32(8),
+				QueueLengthLimit:         int32(65536),
+				Queues:                   int32(128),
+			},
+		},
+	}
+
+	serverConcurrencyLimit := 100
+	requestWaitLimit := 5 * time.Second
+	requests := 5000
+
+	var count int64
+	delegate := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		atomic.AddInt64(&count, int64(1))
+	})
+
+	server, _ := createRequestManagementServerAndClient(
+		t,
+		delegate, flowSchemas, priorityLevelConfigurations,
+		serverConcurrencyLimit, requestWaitLimit,
+	)
+	defer server.Close()
+
+	header := make(http.Header)
+	header.Add("PRIORITY", "0")
+	req, _ := http.NewRequest("GET", server.URL, nil)
+	req.Header = header
+
+	w := &Work{
+		Request: req,
+		N:       requests,
+		// NOTE: C >= serverConcurrencyLimit or else no packets are sent via Work
+		C: 100,
+	}
+	// Run blocks until all work is done.
+	w.Run()
+
+	if count != int64(requests) {
+		t.Errorf("Expected to send %d requests, found %v", requests, count)
+	}
+}
+
+// TestReqMgmtMultiple verifies that fairness is preserved across 5 flowschemas
+// and that the serverConcurrencyLimit/assuredConcurrencyShares are correctly
+// distributed
+func TestReqMgmtMultiple(t *testing.T) {
+	// init objects
+	groups := 5
+	flowSchemas := []*rmtypesv1a1.FlowSchema{}
+	for i := 0; i < groups; i++ {
+		flowSchemas = append(flowSchemas,
+			&rmtypesv1a1.FlowSchema{
+				ObjectMeta: metav1.ObjectMeta{Name: fmt.Sprintf("fs-%d", i)},
+				Spec: rmtypesv1a1.FlowSchemaSpec{
+					PriorityLevelConfiguration: rmtypesv1a1.PriorityLevelConfigurationReference{
+						fmt.Sprintf("plc-%d", i),
+					},
+					MatchingPrecedence: int32(i), // TODO(aaron-prindle) currently ignored
+					// DistinguisherMethod: *rmtypesv1a1.FlowDistinguisherMethodByUserType, // TODO(aaron-prindle) currently ignored
+					// Rules: []rmtypesv1a1.PolicyRuleWithSubjects{}, // TODO(aaron-prindle) currently ignored
+				},
+				// Status: rmtypesv1a1.FlowSchemaStatus{},
+			},
+		)
+	}
+	priorityLevelConfigurations := []*rmtypesv1a1.PriorityLevelConfiguration{}
+
+	for i := 0; i < groups; i++ {
+		priorityLevelConfigurations = append(priorityLevelConfigurations,
+			&rmtypesv1a1.PriorityLevelConfiguration{
+				ObjectMeta: metav1.ObjectMeta{Name: fmt.Sprintf("plc-%d", i)},
+				Spec: rmtypesv1a1.PriorityLevelConfigurationSpec{
+					AssuredConcurrencyShares: 1,
+					Exempt:                   false,
+					GlobalDefault:            false,
+					HandSize:                 int32(8),
+					QueueLengthLimit:         int32(65536),
+					Queues:                   int32(128),
+				},
+			},
+		)
+	}
+
+	serverConcurrencyLimit := groups
+	requestWaitLimit := 5 * time.Second
+	requests := 100
+
+	countnum := len(flowSchemas)
+	counts := make([]int64, len(flowSchemas))
+
+	delegate := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		time.Sleep(500 * time.Millisecond)
```

Comment:
These tests are mixing real time with fake time.  It's not clear that this is correct.

Reply:
This is a good point, I will look more into this and update here.